### PR TITLE
restore dtostrf when floats are disabled in printf/scanf + round fix

### LINKED ISSUE #7043: [BUG]: Rounding Values when Using String Type Cast
### Basic Infos

- [x ] This issue complies with the [issue POLICY doc](https://github.com/esp8266/Arduino/blob/master/POLICY.md).
- [x ] I have read the documentation at [readthedocs](https://arduino-esp8266.readthedocs.io/en/latest) and the issue is not addressed there.
- [x ] I have tested that the issue is present in current master branch (aka latest git).
- [x ] I have searched the issue tracker for a similar issue.
- [ ] If there is a stack dump, I have decoded it.
- [x ] I have filled out all fields below.

#### Platform

- Hardware:			[ESP-12]
- Core Version:			[2.6.3]
- Development Env:		[Arduino IDE|]
- Operating System:		[Windows]

### Settings in IDE

- Module:  			[Generic ESP8266 Module]
- Flash Mode:			[qio|dio|other]
- Flash Size: 		[4MB/1MB]
- lwip Variant:   [v1.4|v2 Lower Memory|Higher Bandwidth]
- Reset Method:		[ck|nodemcu]
- Flash Frequency:	[40Mhz]
- CPU Frequency:		[80Mhz|160MHz]
- Upload Using:		[OTA|SERIAL]
- Upload Speed:   [115200|other] (serial upload only)

### Problem Description

When using a type cast to cast an float to a String like:
`String(pressure, 0)` the result is not rounded correctly.
If I make a type cast like this:
`String(pressure, 1)`the result is rounded correctly.

### [MCVE](https://stackoverflow.com/help/mcve) Sketch

```
#include <Arduino.h>

void setup() {
  Serial.begin(115200);
  while (!Serial) yield(); 
  int a = 1014;
  float b = 1014.45;
  float c = 1014.5;

  Serial.println();
  Serial.println(a);
  
  Serial.println();
  Serial.println(b, 1);
  Serial.println(String(b,1));

  Serial.println();
  Serial.println(c, 0);
  Serial.println(String(c ,0));
}

void loop() {
}
```

### Debug Messages

```
1014

1014.5
1014.5

1015
1014

```

### LINKED ISSUE #7073: dtostrf/Print(float) inconsistent value/methods, downstream use
#7068 removed two slightly different (and both slightly wrong) implementations of `double` to `ASCII string` in `Print::print(float)` and `dtostrf`(a non-std. AVR C addition AFAICT) and replaced it with a call to newlib's `sprintf(%f)` which is known good.

Unfortunately, as @mcspr noted, there are folks who are building the core w/o float-printf support (`printf_float` in platform.txt), so the above PR will end up borking their output.

> Might need to be labelled as breaking? See
https://github.com/arendst/Tasmota/blob/1598e6227a31be5f27081c5a660d22d21613c0ca/platformio.ini#L83
https://github.com/arendst/Tasmota/blob/development/pio/strip-floats.py
https://github.com/xoseperez/espurna/blob/dev/code/scripts/espurna_utils/float_support.py

The main question now is what is the right thing to do in this case.  I suggest it's one of the following:
1) Revert to duplicated code and inconsistent output diffs between Print() and dtostrf() in corner cases.  (i.e. undo #7068).  Larger code for most people (3 different versions of float->ascii in ROM), but downstream keeps working.
2) Keep sprintf(%f) and disallow disabling printf_float, requiring changes in downstream projects (ROM will go up ~1K for them, but down ~1KB for most users since the 2 dup'd versions will be expunged)
3) Pick method from Print(float) and replicate it in dtostrf() so they match.  Then replace Print(float) with a call to dtostrf().  Smaller ROM for all, both match, but may not match the correct output in some cases.
4) Same as 3, but use the dtostrf() as golden and not Print()s version.

FWIW, looking at the AVC-LIBC implementation, it is closest in logic to number 2 above.  They go down to the binary FP level and extract exponents and mantissas (as the AVR ASM level) and work with those parts to actually generate strings.

Any opinions from downstream?

### PR DESCRIPTION
**Pull request by d-a-v**
This PR restores `dtosrf()` only when floats are disabled in printf.
boards.txt builder adds a new define `-DNOPRINTFLOAT` in this case which enables `dtostrf` compilation.
A small fix has been added for the rounding issue, which I believe is generic enough, but will always be a subject for discussion.
Some multiplications have been removed from the algorithm too.

fixes #7073 with solution 6
updates #7068 
fixes #7043 

`dtostrf(1045.5, 0)` now returns `1046`.

@mcspr @arendst Can you test this ?

### CODE REVIEW

**Comment by arendst** [comment]
Did some tests:

- Compiled your PR without changing other parameters results in empty dtostrf responses where normally a valid float would have been printed. Code size 507256 bytes.
- Compiled your PR with additional ``-DNOPRINTFLOAT`` and without pio/strip_floats.py results in valid dtostrf response. Code size 522816. This test doesn't make sense with regard to your PR as floats are enabled anyway.
- Compiled your PR with additional ``-DNOPRINTFLOAT`` and with pio/strip_floats.py results in valid dtostrf response. Code size 507800.

So your PR works but I have to add define ``-DNOPRINTFLOAT`` to get a valid output of dtostrf while still being able to save 15k code space.

**Comment by d-a-v** [comment]
d-a-v approved this pull request.

File: cores/esp8266/core_esp8266_noniso.cpp [hunk 1]
```diff
@@ -40,11 +41,94 @@ char* ultoa(unsigned long value, char* result, int base) {
     return utoa((unsigned int)value, result, base);
 }
 
+#ifdef NOPRINTFLOAT
```
[the remark this answers is not in the document]
Comment (earlephilhower) [reply cores/esp8266/core_esp8266_noniso.cpp]: Dogpiling on this (sorry).  Adding new defines normally ends up breaking PIO or other folks.

Would it be possible, through whatever magic the `-u _printf_float` uses, to somehow make the proper version be linked-in?  

I could also drag this into the `newlib` if that would be needed to make it work "seamlessly" with the existing infra/flags.

[the remark this answers is not in the document]
Comment (mcspr) [reply cores/esp8266/core_esp8266_noniso.cpp]: I thought of it as something like the existing NO_GLOBAL_... defines
PIO user btw, only change that would be required is adding `env.Append(CXXFLAGS=["-DNO_PRINTF_FLOAT"])` to the build script.

Re 'magic', it's just a simple `!= NULL`:
https://github.com/earlephilhower/newlib-xtensa/blob/b326aa447a1e5b33d6ed7d164aa2c59eb3754b9f/newlib/libc/stdio/nano-vfprintf.c#L651-L653

[the remark this answers is not in the document]
Comment (earlephilhower) [reply cores/esp8266/core_esp8266_noniso.cpp]: Oh, then could it be as simple as:

````
char * dtostrf(double number, signed char width, unsigned char prec, char *s) {
    if (!_printf_float) {
        return __dtostrf(number, width, char prec, s);
    } else {
      char fmt[32];
      sprintf(fmt, "%%%d.%df", width, prec);
      sprintf(s, fmt, number);
      return s;
  }
}

static char * __dtostrf(double number, signed char width, unsigned char prec, char *s) {
...all the @d-a-v fixed code here...
}
````

This way the existing flag which will be eval'd a compile time will either cause the dtostrf guts to be omitted and use printf, or vice-versa.

No flag changes, no tool flow changes.  Assuming it works...

[the remark this answers is not in the document]
Comment (d-a-v) [reply cores/esp8266/core_esp8266_noniso.cpp]: great ! :+1: thanks @mcspr @earlephilhower

**Pull request closed without merging**